torch/csrc/jit/tensorexpr/bounds_overlap.h

```diff
@@ -0,0 +1,88 @@
+#pragma once
+
+#include <torch/csrc/jit/tensorexpr/ir_simplifier.h>
+#include <torch/csrc/jit/tensorexpr/ir_visitor.h>
+#include <torch/csrc/jit/tensorexpr/stmt.h>
+#include <deque>
+#include <vector>
+
+namespace torch {
+namespace jit {
+namespace tensorexpr {
+namespace bounds {
+
+// A simple class containing the start and end of a range in a single dimension.
+struct TORCH_API Bound {
+  const Expr* start{nullptr};
+  const Expr* end{nullptr};
+
+  // This stores whether or not the start and end of this Bound have previously
+  // been swapped. This occurs when the bound is in a loop with a negative
+  // stride.
+  bool swapped{false};
+
+  Bound() = default;
+  Bound(const Expr* s, const Expr* e) : start(s), end(e) {}
+
+  void print() const {
+    std::cout << "(" << *start << ", " << *end << ")";
+  }
+
+  bool equals(const Bound& other) const {
+    return exprEquals(start, other.start) && exprEquals(end, other.end);
+  }
+
+  void swap() {
+    std::swap(start, end);
+    swapped = !swapped;
+  }
+};
+
+enum OverlapKind {
+  NoOverlap = 0,
+  PartialOverlap,
+  TotalOverlapA,
+  TotalOverlapB
+};
+
+// Returns the kind of overlap between Bound a and Bound b in a single
+// dimension.
+//     PartialOverlap if any elements in the bound b are in the bound a.
+//     TotalOverlapA if all elements in the bound b are in the bound a.
+//     TotalOverlapB if all elements in the bound a are in the bound b, or if
+//        they are equal.
+//     NoOverlap if no elements in the bound b are in the bound a.
+OverlapKind TORCH_API boundOverlap(Bound a, Bound b);
+
+// A multi dimensional bound representing the bound of a set of indices.
+using IndexBounds = std::vector<Bound>;
+
+// Returns true if two IndexBounds are equivalent.
+bool TORCH_API indexBoundsEquals(const IndexBounds& A, const IndexBounds& B);
+
+// Flattens a multi dimensional bound to a single dimension. The IndexBounds "a"
+// *must* encapsulate the entire range of the buffer.
+Bound TORCH_API flattenBounds(const IndexBounds& a);
+
+// Determines the kind of overlap in X dimensions.
+OverlapKind TORCH_API overlaps(const IndexBounds& a, const IndexBounds& b);
+
+// Returns the bound slices created by subtracing bound B from bound A.
+// Multiple Bounds can be returned in the case where B slices A into two
+// distinct regions with no overlap.
+std::vector<Bound> TORCH_API subtractBound(Bound a, Bound b);
```

Comment:
I chose not to here because while the types match this does not represent indices, but rather separate non-overlapping slices of the same dimension. I will add a comment for this.